### DISCUSSION TITLE
ZbZCLRawReceived & Berry parser

### DISCUSSION THREAD

**Discussion by byka-org** [arendst/Tasmota#16849] [General]
Hi all

Usually when remote zigbee device send some event I can see at debug log ZbZCLReceived, ZbZCLRawReceived and then ZbReceived messages. It's fine.
Using "tasmota.add_rule" on that message we can do something useful with Berry scripting.

But sometimes no ZbReceived message is appears and no MQTT message is sent. I have no idea why.

I look at code and found something like:
#ifdef USE_BERRY
    // Berry pre-process messages
    callBerryZigbeeDispatcher("attributes_raw", &zcl_received, &attr_list, srcaddr);
#endif // USE_BERRY

So, exactly after send ZbZCLRawReceived log message we can parse same data packet with Berry, yeah?
Looking at code: I need to add handler for event "attributes_raw" to class zb_coord - but how to do it?
I found method "add_handler" of class "zb_coord", but how to access to this class?

Is something like "add_handler" command at Berry console, like "add_cmd" or "add_driver"?
I cannot found it at docs or discussions...

**Comment by sfromis** [comment]
What you *can* pick up using `tasmota.add_rule` are only the payloads published to MQTT, and tagged with `MQT:`, *not* the various debugging messages you see when upping the log level to 3, those are not meant as a basis for doing any automation.

**Reply by byka-org** [reply]
Thanks!

I looked deeper at docs and found it:

https://github.com/arendst/Tasmota/pull/16298

It is experimental feature now.
Useful for me because I want to stop using external "big" computers as possible.
